### ISSUE TITLE
Significantly slow training with TFRecords when the labels are one-hot encoded

### ISSUE THREAD

**Issue by sayakpaul** [tensorflow/tensorflow#34435]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Colab
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary):
- TensorFlow version (use command below): 2.0.0
- Python version: 3.6
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10.0
- GPU model and memory: K80

**Describe the current behavior**

I am incorporating TFRecords along with all the `tf.data` legacy functions in my projects. I primarily work with images. I am trying to develop a data input pipeline for an image classification project with the **Flowers-17** dataset. My aim is to first serialize the entire dataset in form TFRecords with multiple shards. I have been able to do that. I am then reading those TFRecords, preprocessing them, augmenting them and finally, I am feeding them to a model. 

Now, I ran two versions of the above experiment:
- One where the image labels were simply encoded using indexing
- Another one where the image labels were one-hot encoded

While compiling the model for the above two scenarios I supplied the loss functions accordingly as well. It now seems that the first experiment where the model **was not** supplied one-hot encoded class labels is training _much faster_ than the other one. 

**Describe the expected behavior**

Model training time should be closely equal in both the cases I mentioned. 

**Code to reproduce the issue**
- [Colab notebook](https://colab.research.google.com/github/sayakpaul/TF-2.0-Hacks/blob/master/tf.data%20exploration/TFRecords_with_tf_data_Advanced.ipynb) for where one-hot encoding wasn't done. 
- [Colab notebook](https://colab.research.google.com/drive/1CuOZJNlWsXdGlhCCfZ14PvSEU2Gtcrdf) with one-hot encoding.

**Comment by oanush** [comment]
@sayakpaul,
Hi,when tried running both the colab gist, I model training time was almost same. Kindly find the [gist](https://colab.sandbox.google.com/gist/oanush/6e9acb4750eb1c1227d4b500015d55a8/copy-of-tfrecords_with_tf_data_advanced-one-hot.ipynb) of colab for one-hot encoding. Thanks!
